Bugfix: Always verify TLS unless explicitly told otherwise

mattbostock opened the pull request:
TLS was not being verified in a number of places:

- connections to grafana.com

- connections to OAuth providers when TLS client authentication was
  enabled

- connections to self-hosted Grafana installations when using the CLI
  tool

TLS should always be verified unless the user explicitly enables an
option to skip verification.

Removes some instances where `InsecureSkipVerify` is explicitly set to
`false`, the default, to help avoid confusion and make it more difficult
to regress on this fix by accident.

Adds a `--insecure` flag to `grafana-cli` to skip TLS verification.

Adds a `tls_skip_verify_insecure` setting for OAuth.

Adds a `app_tls_skip_verify_insecure` setting under a new `[plugins]`
section.

I'm not super happy with the way the global setting is used by
`pkg/api/app_routes.go` but that seems to be the existing pattern used.
16c5d0e  

Fixes #9373, #9419 and is similar to #9377 (which fixes this for datasources).

mattbostock committed:
Always verify TLS unless explicitly told otherwise

TLS was not being verified in a number of places:

- connections to grafana.com

- connections to OAuth providers when TLS client authentication was
  enabled

- connections to self-hosted Grafana installations when using the CLI
  tool

TLS should always be verified unless the user explicitly enables an
option to skip verification.

Removes some instances where `InsecureSkipVerify` is explicitly set to
`false`, the default, to help avoid confusion and make it more difficult
to regress on this fix by accident.

Adds a `--insecure` flag to `grafana-cli` to skip TLS verification.

Adds a `tls_skip_verify_insecure` setting for OAuth.

Adds a `app_tls_skip_verify_insecure` setting under a new `[plugins]`
section.

I'm not super happy with the way the global setting is used by
`pkg/api/app_routes.go` but that seems to be the existing pattern used.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/grafana/grafana?pullRequest=9378) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/grafana/grafana?pullRequest=9378) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/grafana/grafana?pullRequest=9378) it.</sub>

bergquist commented on pkg/api/app_routes.go:
```diff
@@ -18,8 +18,7 @@ import (
 	TLSClientConfig: &tls.Config{
-		InsecureSkipVerify: true,
-		Renegotiation:      tls.RenegotiateFreelyAsClient,
+		Renegotiation: tls.RenegotiateFreelyAsClient,
```
This have to be configurable. In case people want to turn it off.

bergquist replied:
Not sure where this should be configured. I guess we should make a new group 
```
[plugins] 
app_route_skip_tls_verification = false
```

mattbostock replied:
`app_skip_tls_verification` might be easier to understand for a user?

bergquist replied:
your right. `route` doesnt make it clearer. 

This setting would be used for custom routes in app plugins where Grafana proxies the request for the plugin. I would prefer to have an configuration for it.

mattbostock replied:
Ok, thanks for clarifying.

It looks like there's already per-app configuration, so perhaps this setting would be better configured per-app?

mattbostock replied:
One downside to making the setting per-app is you'd need a HTTP transport per app, which probably means more connections since connections will not be re-used between apps.

mattbostock replied:
I'll use a global setting that applies to all app plugins.

bergquist commented on pkg/api/login_oauth.go:
```diff
@@ -97,9 +97,8 @@ func OAuthLogin(ctx *middleware.Context) {
-				InsecureSkipVerify: true,
-				Certificates:       []tls.Certificate{cert},
-				RootCAs:            caCertPool,
+				Certificates: []tls.Certificate{cert},
```
This have to be configurable. In case people want to turn it off.

bergquist replied:
InsecureSkipVerify could be added to the Oauth implementation like TlsClientCert currently works. For some reason its not documented or describe in the config files... but you can find the implementation here: https://github.com/grafana/grafana/blob/master/pkg/social/social.go#L54

bergquist reviewed (changes requested):
This make total sense. But we have to have a setting for disabling this in some cases since grafana is used internal in so many small companies. 

I think we can have one global setting for this except datasources. Datasources should be configured as you suggested in the other PR:

mattbostock commented:
> I think we can have one global setting for this except datasources.

There's a currently a separate setting for at least MySQL, SMTP, LDAP and datasources:
https://github.com/grafana/grafana/search?utf8=%E2%9C%93&q=verify&type=
https://github.com/grafana/grafana/search?utf8=%E2%9C%93&q=SkipVerifySSL&type=

So I think that would mean either making all of those use the single global suggesting, or giving everything its own setting.

What are the expected use cases for skipping verification? The only legitimate one I can think of is for testing/prototyping. The other (less legitimate, IMHO) use case being if you were stuck using a vendor with poor/incorrect/expired TLS.

bergquist commented:
@mattbostock I've seen to many places with terrible TLS setups. While I agree with you that its a bad idea I don't want Grafana to enforce verification but rather enable it by default. 

Enabling easy testing/prototyping as you mentioned is also important. Having one global setting might be a bad idea. I've updated the comments I made earlier on this PR.

mattbostock commented:
Thanks @bergquist, I'll have a look to see what changes would be required to make this a per-feature setting.

mattbostock commented on pkg/api/grafana_com_proxy.go:
```diff
@@ -14,8 +13,7 @@ import (
 var grafanaComProxyTransport = &http.Transport{
-	TLSClientConfig: &tls.Config{InsecureSkipVerify: false},
-	Proxy:           http.ProxyFromEnvironment,
+	Proxy: http.ProxyFromEnvironment,
```
Presumably this doesn't need to be user-configurable as `Grafana.com` should always have a valid certificate.

bergquist replied:
Yes.

mattbostock committed:
OAuth: Check both TLS client cert and key

If either is set, try to use them.

This should help avoid a situation where someone has half-configured TLS
client authentication and it doesn't work without raising an obvious
error.

mattbostock committed:
OAuth: Separate TLS client auth and CA config

It should be specify to either use TLS client authentication or use a
user-supplied CA; previously you had to enable client authentication to
use a custom CA.

mattbostock committed:
OAuth: Rename sslcli

Rename `sslcli` to the more descriptive `oauthClient`.

mattbostock commented:
Update to make all the places we discussed configurable.

I haven't tested these changes beyond running `make test` so any help with testing (particularly from users using these features) would be gratefully appreciated.

mattbostock commented on pkg/setting/setting.go:
```diff
@@ -560,6 +563,9 @@ func NewConfigContext(args *CommandLineArgs) error {
 	authBasic := Cfg.Section("auth.basic")
 	BasicAuthEnabled = authBasic.Key("enabled").MustBool(true)
 
+	// global plugin settings
```
Will fix the capitlisation.

mattbostock replied:
On second thoughts, this is using the same capitalisation as its neighbours so I'll leave it as-is.

bergquist commented:
Looks good. 

I would like to do some testing before merging it. I wonder how hard it would be to write tests for some of these use cases.

bergquist commented:
Thank you for contributing!
